File: android/src/main/java/no/uio/mobileapps/reactnativenettskjema/nettskjema/NettskjemaFormSubmission.java
/*
 * Copyright (c) 2016, University of Oslo, Norway All rights reserved.
 * 
 * This file is part of "UiO Software Information Inventory".
 * 
 * "UiO Software Information Inventory" is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at
 * your option) any later version.
 * 
 * "UiO Software Information Inventory" is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with "UiO Software Information Inventory". If
 * not, see <http://www.gnu.org/licenses/>
 */
package no.uio.mobileapps.reactnativenettskjema.nettskjema;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import no.uio.mobileapps.reactnativenettskjema.interfaces.CsrfRequestFactory;
import no.uio.mobileapps.reactnativenettskjema.interfaces.CsrfTokenFactory;
import no.uio.mobileapps.reactnativenettskjema.interfaces.FilledInFormField;
import no.uio.mobileapps.reactnativenettskjema.interfaces.FormSubmission;
import no.uio.mobileapps.reactnativenettskjema.interfaces.FormSubmissionStatus;
import no.uio.mobileapps.reactnativenettskjema.interfaces.FormSubmissionStatusCode;
import no.uio.mobileapps.reactnativenettskjema.interfaces.SubmissionRequestFactory;
import okhttp3.OkHttpClient;
import okhttp3.Response;

class NettskjemaFormSubmission implements FormSubmission {

    private final OkHttpClient client;
    private final CsrfRequestFactory csrfRequestFactory;
    private final CsrfTokenFactory csrfTokenFactory;
    private final SubmissionRequestFactory postRequestFactory;
    private String responseString;

    public NettskjemaFormSubmission(OkHttpClient client, CsrfRequestFactory csrfRequestFactory, CsrfTokenFactory csrfTokenFactory, SubmissionRequestFactory postRequestFactory) {
        this.client = client;
        this.csrfRequestFactory = csrfRequestFactory;
        this.csrfTokenFactory = csrfTokenFactory;
        this.postRequestFactory = postRequestFactory;
    }

    @Override
    public void post() throws IOException {
        Response csrfResponse = client.newCall(csrfRequestFactory.newRequest()).execute();
        FilledInFormField csrfToken = csrfTokenFactory.newCsrfToken(csrfResponse.body().string());
        Response response = client.newCall(postRequestFactory.newRequest(csrfToken)).execute();
        this.responseString = removeTextAreaTag(response.body().string());
    }

    private static String removeTextAreaTag(String completeStringResponse) {
        return completeStringResponse.substring(10, completeStringResponse.length() - 11);
    }

    @Override
    public FormSubmissionStatus status() {
        Log.d("TEST", responseString);
        if (responseString == null) return new FormSubmissionStatus() {
            @Override
            public FormSubmissionStatusCode statusCode() {
                return FormSubmissionStatusCode.NOT_POSTED;
            }

            @Override
            public String description() {
                return "Submission not yet posted";
            }
        };
        if (isSuccessful()) return new FormSubmissionStatus() {
            @Override
            public FormSubmissionStatusCode statusCode() {
                return FormSubmissionStatusCode.POST_SUCCESSFUL;
            }

            @Override
            public String description() {
                return "Submission successfully posted";
            }
        };
        else return new FormSubmissionStatus() {
            @Override
            public FormSubmissionStatusCode statusCode() {
                return FormSubmissionStatusCode.POST_FAILED;
            }

            @Override
            public String description() {
                    return "Server response: " + responseString;
            }
        };
    }

    private boolean isSuccessful() {
        if (responseString == null) throw new IllegalStateException();
        try {
            JSONObject responseJson = new JSONObject(responseString);
            return responseJson.getString("message").equals("Svaret er levert")
                    && responseJson.getString("status").equals("success");
        } catch (JSONException e) {
            return false;
        }
    }

}

File: ios/ReactNativeNettskjema/ReactNativeNettskjema/NettskjemaFormSubmission.swift
private class NotYetPostedStatus: FormSubmissionStatus {
    let statusCode = FormSubmissionStatusCode.NOT_POSTED
    let description = "Submission not yet posted"
}

private class PostFailedStatus: FormSubmissionStatus {
    let error: String
    
    init(error: String) {
        self.error = error
    }
    
    let statusCode = FormSubmissionStatusCode.POST_FAILED
    var description: String {
        get {
            return "Submission failed with error: " + error
        }
    }
}

private class PostSucceeded: FormSubmissionStatus {
    let statusCode = FormSubmissionStatusCode.POST_SUCCESSFUL
    let description = "Submission successfully posted"
}

class NettskjemaFormSubmission: FormSubmission {
    
    private let csrfRequestFactory: CsrfRequestFactory
    private let csrfTokenFactory: CsrfTokenFactory
    private let filledInForm: FilledInForm
    
    init(sessionManager: String?, csrfRequestFactory: CsrfRequestFactory, csrfTokenFactory: CsrfTokenFactory, filledInForm: FilledInForm) {
        self.csrfRequestFactory = csrfRequestFactory
        self.csrfTokenFactory = csrfTokenFactory
        self.filledInForm = filledInForm
    }
    
    func post(onComplete: (status: FormSubmissionStatus) -> Void) {
        print("TEST: call to post")
        csrfRequestFactory.newRequest({ response in
            print("TEST: in csrf response closure")
            switch response.result {
            case .Success(let data):
                self.filledInForm.postRequest(self.csrfTokenFactory.newCsrfToken(data), onComplete: { encodingResult in
                    switch encodingResult {
                    case .Success(let upload, _, _):
                        print("TEST: upload success")
                        upload.responseString { response in
                            print ("TEST: upload response closure")
                            let stringResponse = String(response)
                            if stringResponse.containsString("failure") {
                                onComplete(status: PostFailedStatus(error: stringResponse))
                            } else {
                                onComplete(status: PostSucceeded())
                            }
                        }
                    case .Failure(let encodingError):
                        print("TEST: upload failed")
                        onComplete(status: PostFailedStatus(error: String(encodingError)))
                    }
                })
            case .Failure(let error):
                print("DBG: csrf request failed")
                onComplete(status: PostFailedStatus(error: error.description))
            }
        })
    }
}
